mesh/amqp.c

```diff
@@ -533,22 +573,53 @@ static bool try_to_connect(struct amqp_thread_context *context)
 	return false;
 }
 
+static inline bool topic_match(const void *a, const void *b)
+{
+	return strcmp(a, b) == 0;
+}
+
+static bool amqp_subscribe_handler(struct message *msg,
+					struct amqp_thread_context *context)
+{
+	char *topic = l_strdup(msg->subscribe.topic);
+
+	if (l_queue_find(context->subscriptions, topic_match, topic)) {
+		l_warn("Exchange topic: '%s' already subscribed!", topic);
+		l_free(topic);
+		return true;
+	}
+
+	if (!l_queue_push_tail(context->subscriptions, topic)) {
+		l_free(topic);
+		return false;
+	}
+
+	if (context->amqp_state == MESH_AMQP_STATE_CONNECTED) {
+		if (!amqp_subscribe_topic(topic, context)) {
+			l_free(topic);
```

Comment:
I think you need to remove `topic` from `context->subscriptions`, otherwise you risk double-freeing it.

Reply:
You are right